src/main/java/com/microsoft/aad/msal4j/TokenCache.java

```diff
@@ -300,16 +302,92 @@ static private AppMetadataCacheEntity createAppMetadataCacheEntity(TokenRequestE
+                        filter(acc -> acc.homeAccountId().contains(acc.localAccountId())).
+                        collect(Collectors.toSet());
+                Set<AccountCacheEntity> tenantAccounts = accounts.values().stream().
+                        filter(acc -> environmentAliases.contains(acc.environment())).
+                        filter(acc -> !acc.homeAccountId().contains(acc.localAccountId())).
+                        collect(Collectors.toSet());
+
+                for (AccountCacheEntity homeAccCached : homeAccounts) {
+                    IdTokenCacheEntity homeAcctIdToken = idTokens.get(getIdTokenKey(
+                            homeAccCached.homeAccountId(),
+                            homeAccCached.environment(),
+                            clientId,
+                            homeAccCached.realm()));
+
+                    Account homeAcc = (Account) homeAccCached.toAccount();
+                    homeAcc.idTokenClaims = JWTParser.parse(homeAcctIdToken.secret).getJWTClaimsSet().getClaims();
+
+                    Iterator<AccountCacheEntity> tenantAcctsIterator = tenantAccounts.iterator();
+                    Map<String, ITenantProfile> tenantProfiles = new HashMap<>();
+
+                    while (tenantAcctsIterator.hasNext()) {
+                        AccountCacheEntity tenantAccCached = tenantAcctsIterator.next();
+                        //If the tenant account's home ID matches a home ID (UID) in the list of home accounts, it is a
+                        //tenant profile of that home account
+                        if (tenantAccCached.homeAccountId.equals(homeAccCached.homeAccountId())) {
+                            IdTokenCacheEntity token = idTokens.get(getIdTokenKey(
+                                    tenantAccCached.homeAccountId(),
+                                    tenantAccCached.environment(),
+                                    clientId,
+                                    tenantAccCached.realm()));
+
+                            ITenantProfile profile = new TenantProfile(tenantAccCached.localAccountId(),
+                                            tenantAccCached.realm(),
+                                            JWTParser.parse(token.secret()).getJWTClaimsSet().getClaims());
+                            tenantProfiles.put(tenantAccCached.realm(), profile);
+                            tenantAcctsIterator.remove();
+                        }
+                    }
+
+                    if (tenantProfiles.size() > 0) {
+                        homeAcc.tenantProfiles(tenantProfiles);
+                    }
+
+                    rootAccounts.add(homeAcc);
+                }
+                //Create Accounts from any tenant profile which did not have a corresponding home account,
+                // and add it to the list of root accounts
+                for (AccountCacheEntity tenantProfileNoHome : tenantAccounts) {
```

Comment:
why do you call accounts tenantProfile?

Reply:
If an AccountCacheEntity is a tenant profile, it's provided to the user by getting attached to a home Account. However, if there is not home account to attach to, then getAccounts() will not be returned them unless they're returned as their own objects. Turning them into Accounts seemed like the correct way to do this, because:
- These AccountCacheEntities would've been returned them as Accounts before this change anyway
- We agreed that TenantProfile should not extend Account, so if I returned them as TenantProfiles I would've had to change the return type of getAccounts(), which would've definitely been a breaking change
- The design doc doesn't really cover this, but this is the behavior done in the Android/iOS libraries, so worst-case at least we're consistent.